node/network/src/protocol.rs

```diff
@@ -1,24 +1,194 @@
+    }
+}
+
+impl ChainStateRetriever for ClientChainStateRetriever {
+    #[inline]
+    fn get_chain_state(&self) -> ChainState {
+        ChainState {
+            genesis_hash: self.client.beacon_chain.chain.genesis_hash(),
+            last_index: self.client.beacon_chain.chain.best_block().index(),
+        }
+    }
+}
+
+/// Protocol responsible for actual message processing from network and sending messages.
+struct Protocol {
+    client: Arc<Client>,
+    peer_manager: Arc<PeerManager<ClientChainStateRetriever>>,
+    inc_gossip_tx: Sender<Gossip<ChainPayload>>,
+    inc_block_tx: Sender<CoupledBlock>,
+}
+
+impl Protocol {
+    fn receive_message(&self, peer_id: PeerId, data: Vec<u8>) {
+        match Decode::decode(&data) {
+            Ok(m) => match m {
+                Message::Connected(connected_info) => {
+                    self.on_new_peer(peer_id, connected_info);
+                },
+                Message::Transaction(tx) => {
+                    if let Err(e) = self.client.shard_client.pool.add_transaction(*tx) {
+                        error!(target: "network", "{}", e);
+                    }
+                },
+                Message::Receipt(receipt) => {
+                    if let Err(e) = self.client.shard_client.pool.add_receipt(*receipt) {
+                        error!(target: "network", "{}", e);
+                    }
+                },
+                Message::Gossip(gossip) => forward_msg(self.inc_gossip_tx.clone(), *gossip),
+                Message::BlockAnnounce(block) => {
+                    forward_msg(self.inc_block_tx.clone(), *block);
+                },
+                Message::BlockRequest(request_id, hashes) => {
+                    match self.client.fetch_blocks(hashes) {
+                        Ok(blocks) => self.send_block_response(&peer_id, request_id, blocks),
+                        Err(err) => {
+                            self.peer_manager.suspect_malicious(&peer_id);
+                            warn!(target: "network", "Failed to fetch blocks from {} with {}. Possible grinding attack.", peer_id, err);
+                        }
+                    }
+                }
+                Message::BlockFetchRequest(request_id, from_index, til_index) => {
+                    match self.client.fetch_blocks_range(from_index, til_index) {
+                        Ok(blocks) => self.send_block_response(&peer_id, request_id, blocks),
+                        Err(err) => {
+                            self.peer_manager.suspect_malicious(&peer_id);
+                            warn!(target: "network", "Failed to fetch blocks from {} with {}. Possible grinding attack.", peer_id, err);
+                        }
+                    }
+                }
+                Message::BlockResponse(_request_id, blocks) => {
+                    for coupled_block in blocks {
+                        forward_msg(self.inc_block_tx.clone(), coupled_block);
```

Comment:
We need to create `forward_msgs` method in which we use `send_all` and `stream::iter_ok` instead of `send` on the channel. Then pass `blocks` to `forward_msgs` instead of spawning a task for each message.